### PR TITLE
Transformers 4 support with transformer 3 checkpoint - addresses #40

### PR DESCRIPTION
**Pull request by cmacdonald**

### CODE REVIEW

**Review by seanmacavaney — COMMENTED** [review]
I don't think showing that it runs on transformers>=4 is sufficient -- don't we also need to show that the embeddings produced are comparable (or give comparable results)?

Hopefully, this isn't a problem between versions 3 and 4. But I know I've seen differences between even earlier versions of the library and version 4.

If there are differences, maybe it's not a showstopper, but we should give a warning that says results may differ from version 3 (and therefore results from the original ColBERT paper, follow-up works, etc.).

File: pyterrier_colbert/__init__.py [hunk 1]
```diff
@@ -27,7 +27,14 @@ def load_checkpoint(path, model, optimizer=None, do_print=True):
 
     checkpoint['model_state_dict'] = new_state_dict
 
-    model.load_state_dict(checkpoint['model_state_dict'])
+    # check for transformer version mismatch and patch address accordingly
+    import transformers
+    from packaging import version
+    strict = True
+    if version.parse(transformers.__version__).major >= 4 and 'bert.embeddings.position_ids' not in checkpoint['model_state_dict']:
```
[the remark this answers is not in the document]
Comment (cmacdonald) [reply pyterrier_colbert/__init__.py:34]: See https://github.com/huggingface/transformers/issues/6882#issuecomment-884730078 - it seems these are correctly initialised regardless.

**Comment by cmacdonald** [comment]
@Xiao0728 can you confirm that performances are unchanged using transformers 4? Y/N?

i think I can add a test that verifies a query encoding matches a tensor computed on transformers 3.

**Comment by Xiao0728** [comment]
Yup, I have tested the transformers with version: 3.0.2 and 4.10.0, and obtain the following results:
image
image
Overall, no clear impact on the performance of the ColBERT e2e model.

**Comment by seanmacavaney** [comment]
Great, thanks for checking. The differences fall within a normal range.

**Comment by cmacdonald** [comment]
thanks everyone

**Pull request merged**